### ISSUE TITLE
Distinguish `ImplStored` when set by the user and `ImplStored` set by heuristic

### ISSUE THREAD

**Issue by inducer** [inducer/pytato#362]
The first should be considered mandatory, for the second, it's OK if array contraction gets rid of it.

**Comment by kaushikcfd** [comment]
We currently don't untag arrays that were already tagged with `ImplStored` in `pt.materialize_with_mpms`.

**Comment by inducer** [comment]
Yeah, but `pt.materialize_with_mpms` could simply use a different tag.
